[CI] IngestDocumentTests testEqualsAndHashcode failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/iuvskxypjqb3w/tests/:server:test/org.elasticsearch.ingest.IngestDocumentTests/testEqualsAndHashcode

**Reproduction line:**
`./gradlew ':server:test' --tests "org.elasticsearch.ingest.IngestDocumentTests.testEqualsAndHashcode" -Dtests.seed=B1AA79F9BA817BBE -Dtests.locale=nl-NL -Dtests.timezone=Africa/Ceuta -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.ingest.IngestDocumentTests&tests.test=testEqualsAndHashcode

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: not <IngestDocument{ sourceAndMetadata={Q=true, bvAOFNKQ=[NXjFOydK, drdCuPnW], Ead=kEtECoGnVP, AiJW=-1572395151, _version_type=external_gte, _version=7482894102100815864}, ingestMetadata={tdDQvC=muTuiZ, VmDQWXOT=RVkvaPXM}}>
     but: was <IngestDocument{ sourceAndMetadata={Q=true, bvAOFNKQ=[NXjFOydK, drdCuPnW], Ead=kEtECoGnVP, AiJW=-1572395151, _version_type=external_gte, _version=7482894102100815864}, ingestMetadata={tdDQvC=muTuiZ, VmDQWXOT=RVkvaPXM}}>

  at __randomizedtesting.SeedInfo.seed([B1AA79F9BA817BBE:C0A5013475663291]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.ingest.IngestDocumentTests.testEqualsAndHashcode(IngestDocumentTests.java:1010)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
+++ b/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
@@ -985,7 +985,10 @@ public class IngestDocumentTests extends ESTestCase {
         if (randomBoolean()) {
             numFields = randomIntBetween(1, IngestDocument.Metadata.values().length);
             for (int i = 0; i < numFields; i++) {
-                Tuple<String, Object> metadata = TestIngestDocument.randomMetadata();
+                Tuple<String, Object> metadata;
+                do {
+                    metadata = TestIngestDocument.randomMetadata();
+                } while (metadata.v2().equals(sourceAndMetadata.get(metadata.v1()))); // must actually be a change
                 otherSourceAndMetadata.put(metadata.v1(), metadata.v2());
             }
             changed = true;

--- a/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
+++ b/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
@@ -17,8 +17,6 @@ import java.time.ZoneId;
 import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
 import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -451,7 +449,7 @@ public class IngestDocumentTests extends ESTestCase {
         @SuppressWarnings("unchecked")
         List<Object> list = (List<Object>) object;
         assertThat(list.size(), equalTo(3));
-        assertThat(list.get(0), equalTo(Collections.singletonMap("field", "value")));
+        assertThat(list.get(0), equalTo(Map.of("field", "value")));
         assertThat(list.get(1), nullValue());
         assertThat(list.get(2), equalTo("new_value"));
     }
@@ -477,13 +475,13 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testListAppendFieldValues() {
-        ingestDocument.appendFieldValue("list", Arrays.asList("item1", "item2", "item3"));
+        ingestDocument.appendFieldValue("list", List.of("item1", "item2", "item3"));
         Object object = ingestDocument.getSourceAndMetadata().get("list");
         assertThat(object, instanceOf(List.class));
         @SuppressWarnings("unchecked")
         List<Object> list = (List<Object>) object;
         assertThat(list.size(), equalTo(5));
-        assertThat(list.get(0), equalTo(Collections.singletonMap("field", "value")));
+        assertThat(list.get(0), equalTo(Map.of("field", "value")));
         assertThat(list.get(1), nullValue());
         assertThat(list.get(2), equalTo("item1"));
         assertThat(list.get(3), equalTo("item2"));
@@ -514,7 +512,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValuesToNonExistingList() {
-        ingestDocument.appendFieldValue("non_existing_list", Arrays.asList("item1", "item2", "item3"));
+        ingestDocument.appendFieldValue("non_existing_list", List.of("item1", "item2", "item3"));
         Object object = ingestDocument.getSourceAndMetadata().get("non_existing_list");
         assertThat(object, instanceOf(List.class));
         @SuppressWarnings("unchecked")
@@ -541,7 +539,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValuesConvertStringToList() {
-        ingestDocument.appendFieldValue("fizz.buzz", Arrays.asList("item1", "item2", "item3"));
+        ingestDocument.appendFieldValue("fizz.buzz", List.of("item1", "item2", "item3"));
         Object object = ingestDocument.getSourceAndMetadata().get("fizz");
         assertThat(object, instanceOf(Map.class));
         @SuppressWarnings("unchecked")
@@ -569,7 +567,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValuesConvertIntegerToList() {
-        ingestDocument.appendFieldValue("int", Arrays.asList(456, 789));
+        ingestDocument.appendFieldValue("int", List.of(456, 789));
         Object object = ingestDocument.getSourceAndMetadata().get("int");
         assertThat(object, instanceOf(List.class));
         @SuppressWarnings("unchecked")
@@ -581,7 +579,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValueConvertMapToList() {
-        ingestDocument.appendFieldValue("fizz", Collections.singletonMap("field", "value"));
+        ingestDocument.appendFieldValue("fizz", Map.of("field", "value"));
         Object object = ingestDocument.getSourceAndMetadata().get("fizz");
         assertThat(object, instanceOf(List.class));
         List<?> list = (List<?>) object;
@@ -590,7 +588,7 @@ public class IngestDocumentTests extends ESTestCase {
         @SuppressWarnings("unchecked")
         Map<String, Object> map = (Map<String, Object>) list.get(0);
         assertThat(map.size(), equalTo(4));
-        assertThat(list.get(1), equalTo(Collections.singletonMap("field", "value")));
+        assertThat(list.get(1), equalTo(Map.of("field", "value")));
     }
 
     public void testAppendFieldValueToNull() {
@@ -628,7 +626,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValuesToListElement() {
-        ingestDocument.appendFieldValue("fizz.list.0", Arrays.asList("item2", "item3", "item4"));
+        ingestDocument.appendFieldValue("fizz.list.0", List.of("item2", "item3", "item4"));
         Object object = ingestDocument.getSourceAndMetadata().get("fizz");
         assertThat(object, instanceOf(Map.class));
         @SuppressWarnings("unchecked")
@@ -674,7 +672,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValuesConvertStringListElementToList() {
-        ingestDocument.appendFieldValue("fizz.list.0.0", Arrays.asList("item2", "item3", "item4"));
+        ingestDocument.appendFieldValue("fizz.list.0.0", List.of("item2", "item3", "item4"));
         Object object = ingestDocument.getSourceAndMetadata().get("fizz");
         assertThat(object, instanceOf(Map.class));
         @SuppressWarnings("unchecked")
@@ -700,7 +698,7 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValueListElementConvertMapToList() {
-        ingestDocument.appendFieldValue("list.0", Collections.singletonMap("item2", "value2"));
+        ingestDocument.appendFieldValue("list.0", Map.of("item2", "value2"));
         Object object = ingestDocument.getSourceAndMetadata().get("list");
         assertThat(object, instanceOf(List.class));
         List<?> list = (List<?>) object;
@@ -709,8 +707,8 @@ public class IngestDocumentTests extends ESTestCase {
         assertThat(list.get(1), nullValue());
         list = (List<?>) list.get(0);
         assertThat(list.size(), equalTo(2));
-        assertThat(list.get(0), equalTo(Collections.singletonMap("field", "value")));
-        assertThat(list.get(1), equalTo(Collections.singletonMap("item2", "value2")));
+        assertThat(list.get(0), equalTo(Map.of("field", "value")));
+        assertThat(list.get(1), equalTo(Map.of("item2", "value2")));
     }
 
     public void testAppendFieldValueToNullListElement() {
@@ -726,15 +724,15 @@ public class IngestDocumentTests extends ESTestCase {
     }
 
     public void testAppendFieldValueToListOfMaps() {
-        ingestDocument.appendFieldValue("list", Collections.singletonMap("item2", "value2"));
+        ingestDocument.appendFieldValue("list", Map.of("item2", "value2"));
         Object object = ingestDocument.getSourceAndMetadata().get("list");
         assertThat(object, instanceOf(List.class));
         @SuppressWarnings("unchecked")
         List<Object> list = (List<Object>) object;
         assertThat(list.size(), equalTo(3));
-        assertThat(list.get(0), equalTo(Collections.singletonMap("field", "value")));
+        assertThat(list.get(0), equalTo(Map.of("field", "value")));
         assertThat(list.get(1), nullValue());
-        assertThat(list.get(2), equalTo(Collections.singletonMap("item2", "value2")));
+        assertThat(list.get(2), equalTo(Map.of("item2", "value2")));
     }
 
     public void testListSetFieldValueIndexProvided() {
@@ -744,7 +742,7 @@ public class IngestDocumentTests extends ESTestCase {
         @SuppressWarnings("unchecked")
         List<Object> list = (List<Object>) object;
         assertThat(list.size(), equalTo(2));
-        assertThat(list.get(0), equalTo(Collections.singletonMap("field", "value")));
+        assertThat(list.get(0), equalTo(Map.of("field", "value")));
         assertThat(list.get(1), equalTo("value"));
     }
 
@@ -755,7 +753,7 @@ public class IngestDocumentTests extends ESTestCase {
         @SuppressWarnings("unchecked")
         List<Object> list = (List<Object>) object;
         assertThat(list.size(), equalTo(2));
-        assertThat(list.get(0), equalTo(Collections.singletonMap("field", "new_value")));
+        assertThat(list.get(0), equalTo(Map.of("field", "new_value")));
         assertThat(list.get(1), nullValue());
     }
 
@@ -1002,7 +1000,7 @@ public class IngestDocumentTests extends ESTestCase {
             }
             changed = true;
         } else {
-            otherIngestMetadata = Collections.unmodifiableMap(ingestMetadata);
+            otherIngestMetadata = Map.copyOf(ingestMetadata);
         }
 
         IngestDocument otherIngestDocument = new IngestDocument(otherSourceAndMetadata, otherIngestMetadata);
@@ -1013,10 +1011,7 @@ public class IngestDocumentTests extends ESTestCase {
             assertThat(ingestDocument, equalTo(otherIngestDocument));
             assertThat(otherIngestDocument, equalTo(ingestDocument));
             assertThat(ingestDocument.hashCode(), equalTo(otherIngestDocument.hashCode()));
-            IngestDocument thirdIngestDocument = new IngestDocument(
-                Collections.unmodifiableMap(sourceAndMetadata),
-                Collections.unmodifiableMap(ingestMetadata)
-            );
+            IngestDocument thirdIngestDocument = new IngestDocument(Map.copyOf(sourceAndMetadata), Map.copyOf(ingestMetadata));
             assertThat(thirdIngestDocument, equalTo(ingestDocument));
             assertThat(ingestDocument, equalTo(thirdIngestDocument));
             assertThat(ingestDocument.hashCode(), equalTo(thirdIngestDocument.hashCode()));

